### ISSUE TITLE
No detection of missing dataclass initialisation arguments

### ISSUE THREAD

**Issue by DavidS-cloud** [python/mypy#13332]
**Bug Report**

Invalid usage of a `@dataclass` class in a function is not reported by mypy.

**To Reproduce**

Created a new venv:

```
❯ python --version
Python 3.10.5
❯ pip list
Package           Version
----------------- -------
autopep8          1.6.0
mypy              0.971
mypy-extensions   0.4.3
pip               22.2
pycodestyle       2.9.1
setuptools        59.6.0
toml              0.10.2
tomli             2.0.1
typing_extensions 4.3.0
```

source:
```
from dataclasses import dataclass


@dataclass
class TestClass:
    a: str
    b: str


def fn():
    test = TestClass(a="a")

    print(test)


fn()
```

test execution:

```
❯ mypy src/test.py && echo ---------  && python3 src/test.py 
Success: no issues found in 1 source file
---------
Traceback (most recent call last):
  File "/home/dschmitt/tmp/mupu/src/test.py", line 16, in <module>
    fn()
  File "/home/dschmitt/tmp/mupu/src/test.py", line 11, in fn
    test = TestClass(a="a")
TypeError: TestClass.__init__() missing 1 required positional argument: 'b'

❯ 
```

**Expected Behavior**

A slight modification of that file correctly fails type checking:

```
from dataclasses import dataclass


@dataclass
class TestClass:
    a: str
    b: str


test = TestClass(a="a")

print(test)
```

```
❯ mypy src/test.py && echo ---------  && python3 src/test.py 
src/test.py:10: error: Missing positional argument "b" in call to "TestClass"
Found 1 error in 1 file (checked 1 source file)

❯ 
```

**Comment by DavidS-cloud** [comment]
oh, it's because `fn()` is not annotated. adding `-> None` to the function correctly causes the error to be flagged.
